### ISSUE TITLE
Valgrind reports memory errors in `profile.jl` test

### ISSUE THREAD

**Issue by garrison** [JuliaLang/julia#10220]
I build `julia` with `-DMEMDEBUG and -DMEMDEBUG2` and run the following from `julia/test/`

    valgrind --smc-check=all-non-file --trace-children=yes --suppressions=$PWD/../contrib/valgrind-julia.supp ../julia runtests.jl profile

and the following errors are reported repeatedly:

    ==771== Conditional jump or move depends on uninitialised value(s)
    ==771==    at 0x86321EA: ???
    ==771==    by 0x86314C8: ???
    ==771==    by 0x862FDC7: ???
    ==771==    by 0x8623CBD: ???
    ==771==    by 0x8632E90: ???
    ==771==    by 0x8632CDA: ???
    ==771==    by 0x4F102E5: jl_apply (julia.h:1071)
    ==771==    by 0x4F141ED: jl_trampoline (builtins.c:934)
    ==771==    by 0x4F102E5: jl_apply (julia.h:1071)
    ==771==    by 0x4F125C2: jl_f_kwcall (builtins.c:505)
    ==771==    by 0x862387C: ???
    ==771==    by 0x8632C1D: ???
    ==771== 
    ==771== Use of uninitialised value of size 8
    ==771==    at 0x86321F4: ???
    ==771==    by 0x86314C8: ???
    ==771==    by 0x862FDC7: ???
    ==771==    by 0x8623CBD: ???
    ==771==    by 0x8632E90: ???
    ==771==    by 0x8632CDA: ???
    ==771==    by 0x4F102E5: jl_apply (julia.h:1071)
    ==771==    by 0x4F141ED: jl_trampoline (builtins.c:934)
    ==771==    by 0x4F102E5: jl_apply (julia.h:1071)
    ==771==    by 0x4F125C2: jl_f_kwcall (builtins.c:505)
    ==771==    by 0x862387C: ???
    ==771==    by 0x8632C1D: ???
    ==771== 

`versioninfo` is as follows:

    Commit f8976b6* (2015-02-17 02:42 UTC)
    DEBUG build
    Platform Info:
      System: Linux (x86_64-linux-gnu)
      CPU: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
      WORD_SIZE: 64
      BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
      LAPACK: libopenblas
      LIBM: libopenlibm
      LLVM: libLLVM-3.3

**Comment by garrison** [comment]
I can reproduce this in d7a90c0, which is one of the earliest commits in which this test existed.  So this problem is at least as old as the test.

**Comment by timholy** [comment]
Thanks for the report. I've spent some time looking into this, and I've come to the conclusion that it's a spurious issue that's nevertheless worth fixing. I think the issue is that valgrind is confused by the fact that `bt_data_prof` is filled from a signal-handler.

I've been able to replicate it with a test as minimal as
```julia
Profile.clear()
@profile busywait(1, 20)
bt = Profile.fetch()
@show Profile.lookup(bt[1])
@show Profile.lookup(convert(Ptr{Void}, 0))
@show length(bt)
n = length(bt)
pglen = 4096
m = floor(Int, pglen/sizeof(UInt))-1
@show m
@show bt[m]
```
(essentially, just reading a few elements of the collected backtrace data) although it's hit-and-miss. AFAICT it requires that last line, but perhaps it would have eventually issued a warning for `bt[1]` if I had tested it enough times. The whole `pglen` thing was me exploring various hypotheses, and I don't think there's anything particularly special about that value of `m`.

On my machine it flags [this line](https://github.com/JuliaLang/julia/blob/233ebd85552345bc550d112480688680d7b0d8f5/src/profile.c#L433) as the source of the problem. If I change it from `malloc` to `calloc`, the problem goes away. I'll submit a patch to do that, mostly to silence this warning, but it probably makes things a little more robust anyway (because we use zeros to indicate end-of-backtrace).
